### ISSUE TITLE
charts: stop using release name in chart templates

### ISSUE THREAD

**Issue by ableuler** [SwissDataScienceCenter/renku#1481]
One annoying thing about helm charts is the fact that values files can not be templated. For our purpose this would be useful when referring to a k8s resource in the values which is itself defined in another chart template (like a secret). 
We currently include the release name in the names of our generated resources (except for things like jupyterhub or postgres where this is not the case). This makes the resource names unpredictable. This is actually one reason that made @Panaetius explore helm alternatives. 
With helm 3 respecting k8s namespaces even more, I think it would be smart to remove the release name from all our templates, get predictable resource names and thus free us from any need to provide `andreas-renku-some-secret` type of values when deploying RenkuLab.
The only drawback I see is that this would make it impossible to have multiple RenkuLab deployments in one k8s namespace, but that's already not really possible now (Jupyterhub), and I also don't see why anybody would want to do that.

What do others think about this proposition? @rokroskar @pameladelgado @Panaetius

**Comment by rokroskar** [comment]
yeah since the expectation is that everything is namespaced then this should be fine. It also just makes the object names super ugly.

**Comment by ableuler** [comment]
Ok, I think I'll take the effort relatively soon then and rip the release names out of our charts. Do people prefer to keep "renku" in the resource names or not even that, for example`renku-keycloak` vs `keycloak`?

**Comment by rokroskar** [comment]
I'm going a bit back and forth on this - It might be worth trying to understand why that is the default in helm. I can see several reasons and it's also how most charts we use behave (with the exception of jupyterhub). Though I guess you would only find this annoying is if you were looking at resource listings across several namespaces, which I personally never really do but maybe if resources were split across namespaces in the future it would make a difference. Addressing services via in-cluster DNS becomes a bit easier since you now would have `<service-name>.<namespace>` instead of `<release>-<service-name>.<namespace>`.

**Comment by ableuler** [comment]
Ha, I yesterday had similar thoughts when realizing that Jupyterhub is trying to go the other way: https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/1740. It indeed seems to be a chart best practice to make several releases into the same namespace possible. On the other hand, unpredictable resource names make it impossible to point to resources (in practice it's always `existingSecret`s) in values files. So what do we want. Multiple RenkuLabs in one namespace or simpler values files?

**Comment by rokroskar** [comment]
Well we're going to need a script in the end to generate the values file anyway. At that point it doesn't really matter so much if we automatically fill in the secret names.

**Comment by ableuler** [comment]
I don't fully agree. As long as we provide a helm chart for RenkuLab (which I am convinced that we have to), people will need to interact with the values file. Therefore, having an as-small-as-possible de-duplicated values file has - well - a lot of value. Providing a script that helps people create an initial values file will definitely improve the situation and get people up and running more quickly, but we should still aim for simpler vaules files whenever possible IMHO.

**Comment by rokroskar** [comment]
certainly - are the secrets that can't be templated needed for a minimal poc? or are they needed just for e.g. enabling TLS?

**Comment by ableuler** [comment]
> or are they needed just for e.g. enabling TLS?
No, it's everywhere. The pattern of "existing secrets" (where a templateable secret name in the values fiel would be useful) is being used in more and more charts (keycloak, postgres, JH more recently). The idea is that you allow secrets to be defined in two ways, either by putting the actual secret into the values file, or by providing the name of an existing k8s secret which contains the actual secret.
This is great for our purpose, because it allows us to create secrets at the renku chart level and tell the dependency charts (for example Jupyterhub) to use it. This allows us omit all the secrets in the Jupyterhub block of the values. Bur for this to work we need to have predictable resource names at the renku chart level.

**Comment by rokroskar** [comment]
can you use the `_helpers.yaml` to define this? We use it for setting some of the names already.

**Comment by ableuler** [comment]
I don't think so. These problems concerned with resource names come from the dependency charts that we do not control. If they expect the name of a k8s passed in though the values file, there's not much we can do about it. Or am I missing something fundamental here?

**Comment by rokroskar** [comment]
Well I wondered if you can make a resource with `<release-name>-<name>` i.e. the pattern we have now but then map that using the helpers to something generic. So you would have the secret object called `andreas-renku-secret` and the helper for that release would map `andreas-renku-secret` to `renku-secret`.

**Comment by kristofmartens** [comment]
I am currently working on a PR that is actually the opposite of what is proposed here. The reason why I am working on this is because we have a use case that requires multiple jupyterhub installations in the same namespace. Right now there are many hardcoded assumptions made between "kubespawner" and the helm templates and this leads to many resource name conflicts. I have a fully operational (0.9.0 based) branch up and running as we speak and am doing some operational tests before I will start the PR on master. If the conclusion will be that release names will not be added (which is against helm recommendations) I will abandon the PR. Please advise on what to do here as there are clearly 2 schools of thought here ;)

**Comment by rokroskar** [comment]
Hi @kristofmartens thanks for weighing in on the discussion! When you say "many hardcoded assumptions" are you referring to renku specifically or more generally in the jupyterhub charts? 

I am, in principle, in favor of following what is typically "officially" endorsed because it leads to fewer surprises for people who are deploying for the first time. I still find the JH charts a bit strange after several years of using them :) 

I'm curious what sort of use-case you have that requires several JH instances in the same namespace?

**Comment by kristofmartens** [comment]
The idea behind needing to have multiple jupyterhubs per namespace is that for security/usability/audit reasons we chose for a model of deploying a jupyterhub instance per project. This would allow us to manage data access on a per project level instead of an individual level to avoid the use of data for a cause where it is not granted for. We combine this per project concept with providing project specific cloud credentials to the single user pods. 
This way data scientists can work on different projects just by accessing a different (authenticated) url and the right access to data is already set for them automatically.

**Comment by ableuler** [comment]
Hi @kristofmartens and welcome to the Renku project. 

> If the conclusion will be that release names will not be added (which is against helm recommendations) I will abandon the PR. Please advise on what to do here as there are clearly 2 schools of thought here ;)

Please note that Renku and Jupyterhub are two independent projects. In Renku we heavily rely und Jupyterhub as a dependency, but we are not maintaining Jupyterhub and therefore can't tell you if your PR will be considered or not. 

Now for the actual topic: Although I don't understand yet why you can't create a k8s namespace for each project/jupyterhub instance, I totally agree that it's helm best practice to add the release name to the resource names. And as a "heavy user" I surely wouldn't mind Jupyterhub going that way. 

The main reason why I suggested that our Renku project might go the other way is how charts (like Renku) interface with their chart dependencies (like Jupyterhub). Specifically, I would like to create a secret in the renku chart and provide it to Jupyterhub through the `existingSecret` value. Since the values files themselves are not templateable, we either need to have a predictable secret name (which does *not* include the release name), or the user needs to fill in the correct secret name (which includes the release name) through the values file.

**Comment by ableuler** [comment]
> Well I wondered if you can make a resource with `<release-name>-<name>` i.e. the pattern we have now but then map that using the helpers to something generic. So you would have the secret object called `andreas-renku-secret` and the helper for that release would map `andreas-renku-secret` to `renku-secret`.

I think this would work if you control the subchart. But if the subchart is off-the-shelf, than there's not way to make the subchart use the helper function. Or maybe I'm not understanding your suggestion correctly?

**Comment by rokroskar** [comment]
You want the subchart to consume a resource called `resource` but your main chart is naming that resource `<release>-resource`. You can use a helper in the main chart to map `<release>-resource` to `resource`, which in turn will make the subchart see it in the way that it expects.

**Comment by rokroskar** [comment]
But I guess this might only apply to the templating engine... 🤔

**Comment by ableuler** [comment]
I think I get your idea of a helper that does that mapping, but as far as I understand there is no way to make the templating engine apply mapping defined at the renku or renku-notebooks level when evaluating [this line](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/079fe5a9f800543dd21f36beae64697125f6c4c5/jupyterhub/templates/hub/deployment.yaml#L45) I think it will just use whatever is written in 
```
notebooks:
  jupyterhub: 
    hub:
      existingSecret: <name-of-the-k8s-resource>
```
